File: vectadd.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

#include <CL/cl.h>

const char* programSource = 
  "__kernel void vectAddInt(__global int *A,\n"
  "                         __global int *B,\n"
  "                         __global int *C,\n"
  "                         int lenbase)\n"
  "{\n"
  "   uint gid = get_global_id(0);\n"
  "   uint gsize = get_global_size(0);\n"
  "\n"
  "   for (uint idx = gid; idx < lenbase; idx += gsize) {\n"
  "     C[idx] = A[idx] + B[idx];\n"
  "   }\n"
  "}\n";

int main() {
  const int elements = 204800000;

  cl_int status;

  cl_uint numPlatforms = 0;
  cl_platform_id *platforms = NULL;

  status = clGetPlatformIDs(0, NULL, &numPlatforms);

  platforms = (cl_platform_id*)malloc(numPlatforms*sizeof(cl_platform_id));

  status = clGetPlatformIDs(numPlatforms, platforms, NULL);

  for (int p = 0; p < numPlatforms; p++) {

    char *pname = NULL;
    size_t psize;
    clGetPlatformInfo(platforms[p], CL_PLATFORM_NAME, psize, pname, &psize);
    pname = (char*)malloc(psize);
    clGetPlatformInfo(platforms[p], CL_PLATFORM_NAME, psize, pname, NULL);

    printf("Name: %s\n", pname);

    char *pvendor = NULL;
    clGetPlatformInfo(platforms[p], CL_PLATFORM_VENDOR, psize, pvendor, &psize);
    pvendor = (char*)malloc(psize);
    clGetPlatformInfo(platforms[p], CL_PLATFORM_VENDOR, psize, pvendor, NULL);

    printf("Vendor: %s\n", pvendor);

    cl_uint numDevices = 0;
    cl_device_id *devices = NULL;

    status = clGetDeviceIDs(platforms[p], CL_DEVICE_TYPE_ALL, 0, NULL, &numDevices);

    devices = (cl_device_id*)malloc(numDevices*sizeof(cl_device_id));

    status = clGetDeviceIDs(platforms[p], CL_DEVICE_TYPE_ALL, numDevices, devices, NULL);

    cl_context context = NULL;
    context = clCreateContext(NULL, numDevices, devices, NULL, NULL, &status);

    for (int d = 0; d < numDevices; d++) {

      char *dname = NULL;
      size_t dsize;
      clGetDeviceInfo(devices[d], CL_DEVICE_NAME, dsize, dname, &dsize);
      dname = (char*)malloc(dsize);
      clGetDeviceInfo(devices[d], CL_DEVICE_NAME, dsize, dname, NULL);

      printf("Name: %s\n", dname);

      char *dvendor = NULL;
      clGetDeviceInfo(devices[d], CL_DEVICE_VENDOR, dsize, dvendor, &dsize);
      dvendor = (char*)malloc(dsize);
      clGetDeviceInfo(devices[d], CL_DEVICE_VENDOR, dsize, dvendor, NULL);

      printf("Vendor: %s\n", dvendor);

      cl_device_type dtype;
      clGetDeviceInfo(devices[d], CL_DEVICE_TYPE, sizeof(cl_device_type), &dtype, NULL);

      switch (dtype) {
      case CL_DEVICE_TYPE_CPU:
	printf("Type: CPU\n");
	break;
      case CL_DEVICE_TYPE_GPU:
	printf("Type: GPU\n");
	break;
      }

      cl_program program = clCreateProgramWithSource(context, 1, (const char**)&programSource, NULL, &status);
  
      status = clBuildProgram(program, numDevices, devices, NULL, NULL, NULL);

      cl_kernel kernel = NULL;

      kernel = clCreateKernel(program, "vectAddInt", &status);

      cl_command_queue cmdQueue;
      cmdQueue = clCreateCommandQueue(context, devices[d], 0, &status);

      // local_work_size
      // * must be evenly divisible by global_work_size (working it backwards here)
      // * can't clash with __attribute__((reqd_work_group_size(X, Y, Z)))
      // * total number of items across dimensions is greater than CL_DEVICE_MAX_WORK_GROUP_SIZE
      // * items in any dimension exceeed CL_DEVICE_MAX_WORK_ITEM_SIZES in that dimension

      size_t local_work_size;
      local_work_size = elements;

      size_t max_work_group_size;
      clGetDeviceInfo(devices[d], CL_DEVICE_MAX_WORK_GROUP_SIZE, dsize, &max_work_group_size, NULL);
      printf("max work group size: %d\n", (int)max_work_group_size);

      if (local_work_size > max_work_group_size) {
	local_work_size = max_work_group_size;
      }

      size_t max_work_item_dimensions;
      clGetDeviceInfo(devices[d], CL_DEVICE_MAX_WORK_ITEM_DIMENSIONS, dsize, &max_work_item_dimensions, NULL);
      printf("max work item dimensions: %d\n", (int)max_work_item_dimensions);

      size_t max_work_item_sizes[3];
      clGetDeviceInfo(devices[d], CL_DEVICE_MAX_WORK_ITEM_SIZES, sizeof(max_work_item_sizes), &max_work_item_sizes, NULL);
      printf("max work item sizes[0]: %d\n", (int)max_work_item_sizes[0]);

      if (local_work_size > max_work_item_sizes[0]) {
	local_work_size = max_work_item_sizes[0];
      }

      size_t kernel_work_group_size;
      clGetKernelWorkGroupInfo(kernel, devices[d], CL_KERNEL_WORK_GROUP_SIZE, sizeof(size_t), &kernel_work_group_size, NULL);
      printf("kernel work group size: %d\n", (int)kernel_work_group_size);

      if (local_work_size > kernel_work_group_size) {
	local_work_size = kernel_work_group_size;
      }

      size_t kernel_preferred_work_group_size_multiple;
      clGetKernelWorkGroupInfo(kernel, devices[d], CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE, sizeof(size_t), &kernel_preferred_work_group_size_multiple, NULL);
      printf("kernel preferred work group size multiple: %d\n", (int)kernel_preferred_work_group_size_multiple);

      local_work_size = (local_work_size/kernel_preferred_work_group_size_multiple)*kernel_preferred_work_group_size_multiple;

      printf("local work size: %d\n", (int)local_work_size);

      // global_work_size 
      // * can't be bigger than 2^(address_bits)-1 (hah)
      // * must be a multiple of local_work_size
      // * largest memory allocation cannot exceed CL_DEVICE_MAX_MEM_ALLOC_SIZE
      // * total memory allocation (dynamic plus static) cannot exceed CL_DEVICE_GLOBAL_MEM_SIZE

      size_t global_work_size;
      global_work_size = elements;

      int static_data, bpe_single, bpe_total;
      // lenbase is the only static element
      static_data = sizeof(int);
      // size of the largest single element
      bpe_single = sizeof(int);
      // total size of one element's allocation
      bpe_total = 3 * bpe_single;
      
      cl_ulong global_mem_size;
      clGetDeviceInfo(devices[d], CL_DEVICE_GLOBAL_MEM_SIZE, sizeof(cl_ulong), &global_mem_size, NULL);

      if (dtype == CL_DEVICE_TYPE_GPU) {
	// assume less than half the GPU memory is really available
	global_mem_size = (size_t) (0.35 * (float) global_mem_size);
      }
      printf("global mem size: %ld\n", (long)global_mem_size);

      long max_total;
      max_total = (global_mem_size-static_data)/bpe_total;

      if (global_work_size > max_total) {
	global_work_size = max_total;
      }

      cl_ulong max_mem_alloc_size;
      clGetDeviceInfo(devices[d], CL_DEVICE_MAX_MEM_ALLOC_SIZE, sizeof(cl_ulong), &max_mem_alloc_size, NULL);
      printf("max mem alloc size: %ld\n", (long)max_mem_alloc_size);

      long max_single;
      max_single = max_mem_alloc_size/bpe_single;

      if (global_work_size > max_single) {
	global_work_size = max_single;
      }

      global_work_size = (global_work_size/local_work_size)*local_work_size;

      // JMT: testing
      printf("global work size: %d\n", (int)global_work_size);

      cl_uint max_compute_units;
      clGetDeviceInfo(devices[d], CL_DEVICE_MAX_COMPUTE_UNITS, dsize, &max_compute_units, NULL);

      printf("max compute units: %d\n", max_compute_units);

      long multiplier;
      multiplier = global_work_size/(local_work_size*max_compute_units);
      printf("multiplier: %d\n", (int) multiplier);

      int *A = NULL;
      int *B = NULL;
      int *C = NULL;

      size_t datasize = sizeof(int)*elements;

      A = (int*)malloc(datasize);
      B = (int*)malloc(datasize);
      C = (int*)malloc(datasize);

      for (int i = 0; i < elements; i++) {
	A[i] = i;
	B[i] = i;
      }

      // break arrays into sub arrays of 'global_work_size' size
      int Cindex = 0;
      while (Cindex < elements) {
	int sublen;
	sublen = global_work_size;
	if ((elements-Cindex) < global_work_size) {
	  sublen = (elements-Cindex);
	}

	printf("sublen: %d\n", sublen);

	int *subA = NULL;
	int *subB = NULL;
	int *subC = NULL;

	size_t bufsize = sizeof(int)*sublen;
	printf("bufsize: %d\n", (int)bufsize);

	subA = (int*)malloc(bufsize);
	subB = (int*)malloc(bufsize);
	subC = (int*)malloc(bufsize);

	for (int i = 0; i < sublen; i++) {
	  subA[i] = A[Cindex+i];
	  subB[i] = B[Cindex+i];
	}

	cl_mem bufferA;
	cl_mem bufferB;
	cl_mem bufferC;

	bufferA = clCreateBuffer(context, CL_MEM_READ_ONLY, bufsize, NULL, &status);
	if (status != CL_SUCCESS) {
	  printf("create A freak out %d!\n", status);
	}
	bufferB = clCreateBuffer(context, CL_MEM_READ_ONLY, bufsize, NULL, &status);
	if (status != CL_SUCCESS) {
	  printf("create B freak out %d!\n", status);
	}
	bufferC = clCreateBuffer(context, CL_MEM_WRITE_ONLY, bufsize, NULL, &status);
	if (status != CL_SUCCESS) {
	  printf("create C freak out %d!\n", status);
	}

	status = clEnqueueWriteBuffer(cmdQueue, bufferA, CL_TRUE, 0, bufsize, subA, 0, NULL, NULL);
	if (status != CL_SUCCESS) {
	  printf("write A freak out %d!\n", status);
	}

	status = clFinish(cmdQueue);
	if (status != CL_SUCCESS) {
	  printf("write B freak out %d!\n", status);
	}

	status = clEnqueueWriteBuffer(cmdQueue, bufferB, CL_TRUE, 0, bufsize, subB, 0, NULL, NULL);
	if (status != CL_SUCCESS) {
	  printf("write B freak out %d!\n", status);
	}

	status = clFinish(cmdQueue);
	if (status != CL_SUCCESS) {
	  printf("write B freak out %d!\n", status);
	}

	status = clSetKernelArg(kernel, 0, sizeof(cl_mem), &bufferA);
	if (status != CL_SUCCESS) {
	  printf("set A freak out %d!\n", status);
	}
	status |= clSetKernelArg(kernel, 1, sizeof(cl_mem), &bufferB);
	if (status != CL_SUCCESS) {
	  printf("set B freak out %d!\n", status);
	}
	status |= clSetKernelArg(kernel, 2, sizeof(cl_mem), &bufferC);
	if (status != CL_SUCCESS) {
	  printf("set C freak out %d!\n", status);
	}
	status |= clSetKernelArg(kernel, 3, sizeof(cl_int), &sublen);
	if (status != CL_SUCCESS) {
	  printf("set sublen freak out %d!\n", status);
	}

	size_t globalWorkSize[1];
	size_t localWorkSize[1];

	globalWorkSize[0] = sublen;
	localWorkSize[0] = local_work_size;

	status = clEnqueueNDRangeKernel(cmdQueue, kernel, 1, NULL, globalWorkSize, localWorkSize, 0, NULL, NULL);
	if (status != CL_SUCCESS) {
	  printf("kernel freak out %d!\n", status);
	}

	status = clEnqueueReadBuffer(cmdQueue, bufferC, CL_TRUE, 0, bufsize, subC, 0, NULL, NULL);
	if (status != CL_SUCCESS) {
	  printf("read C freak out %d!\n", status);
	}

	// iterate over the output and add it to C
	for (int i = 0; i < sublen; i++) {
	  C[Cindex++] = subC[i];
	}

	clReleaseMemObject(bufferA);
	clReleaseMemObject(bufferB);
	clReleaseMemObject(bufferC);
	free(subA);
	free(subB);
	free(subC);
      } // inside iteration

      bool result = true;
      for (int i = 0; i < elements; i++) {
	if (C[i] != i+i) {
	  printf("C[%d] %d != %d\n", i, C[i], i+i);
	  result = false;
	  break;
	}
      }
      if (result) {
	printf("Output is correct\n");
      } else {
	printf("Output is incorrect\n");
	exit(-1);
      }

      free(A);
      free(B);
      free(C);
      clReleaseKernel(kernel);
      clReleaseProgram(program);
      clReleaseCommandQueue(cmdQueue);
      free(dname);
      free(dvendor);
    } // inside devices
    clReleaseContext(context);
    free(devices);
    free(pname);
    free(pvendor);
  } // inside platforms
  free(platforms);
}

File: vectadd.go
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"log"
	"strconv"
	"strings"

	"github.com/mathuin/go-opencl/cl"
)

func main() {

	const elements = 204800011

	for _, platform := range cl.Platforms {
		log.Printf("Platform: %s", platform.Property(cl.PLATFORM_NAME))
		for _, dev := range platform.Devices {
			log.Printf("Device: %s (%s)", dev.Property(cl.DEVICE_NAME), dev.Property(cl.DEVICE_TYPE))
			log.Printf("Version: %s", dev.Property(cl.DEVICE_VERSION))

			var err error
			var context *cl.Context
			var queue *cl.CommandQueue
			var program *cl.Program
			var kernel *cl.Kernel

			if context, err = cl.NewContextOfDevices(nil, []cl.Device{dev}); err != nil {
				panic(err)
			}

			if queue, err = context.NewCommandQueue(dev, cl.QUEUE_NIL); err != nil {
				panic(err)
			}

			if program, err = context.NewProgramFromFile("vector2.cl"); err != nil {
				panic(err)
			}

			if err = program.Build([]cl.Device{dev}, ""); err != nil {
				if status := program.BuildStatus(dev); status != cl.BUILD_SUCCESS {
					panic(fmt.Errorf("Build Error:\n%s\n", program.Property(dev, cl.BUILD_LOG)))
				}
				panic(err)
			}

			if kernel, err = program.NewKernelNamed("vectAddInt"); err != nil {
				panic(err)
			}

			// JMT: testing new kernel functions
			var kernelNumArgs int
			if kernelNumArgs, err = strconv.Atoi(fmt.Sprintf("%d", kernel.Property(cl.KERNEL_NUM_ARGS))); err != nil {
				panic(err)
			}
			log.Printf("kernel num args: %d", kernelNumArgs)

			var localWorkSize int
			if localWorkSize, err = getLocalWorkSize(dev, kernel, elements); err != nil {
				panic(err)
			}

			// this project has one int32 of static data
			staticData := 4
			// each base element is two int32's
			bpeSingle := 2 * 4
			// each retval is one int32
			bpeTotal := bpeSingle + 4

			// global work size
			var globalWorkSize int
			if globalWorkSize, err = getGlobalWorkSize(dev, kernel, elements, staticData, bpeSingle, bpeTotal, localWorkSize); err != nil {
				panic(err)
			}

			var maxComputeUnits int
			if maxComputeUnits, err = strconv.Atoi(fmt.Sprintf("%d", dev.Property(cl.DEVICE_MAX_COMPUTE_UNITS))); err != nil {
				panic(err)
			}
			multiplier := int(float64(globalWorkSize) / (float64(localWorkSize) * float64(maxComputeUnits)))

			log.Printf("multiplier: %d", multiplier)

			A := make([]int32, elements)
			B := make([]int32, elements)
			C := make([]int32, elements)
			for i := int32(0); i < elements; i++ {
				A[i], B[i] = i, i
			}

			var bufA, bufB, bufC *cl.Buffer
			if bufA, err = context.NewBuffer(cl.MEM_READ_ONLY, uint32(globalWorkSize*4)); err != nil {
				panic(err)
			}

			if bufB, err = context.NewBuffer(cl.MEM_READ_ONLY, uint32(globalWorkSize*4)); err != nil {
				panic(err)
			}

			if bufC, err = context.NewBuffer(cl.MEM_WRITE_ONLY, uint32(globalWorkSize*4)); err != nil {
				panic(err)
			}

			if err = kernel.SetArg(0, bufA); err != nil {
				panic(err)
			}

			if err = queue.Finish(); err != nil {
				panic(err)
			}

			if err = kernel.SetArg(1, bufB); err != nil {
				panic(err)
			}

			if err = queue.Finish(); err != nil {
				panic(err)
			}

			if err = kernel.SetArg(2, bufC); err != nil {
				panic(err)
			}

			if err = queue.Finish(); err != nil {
				panic(err)
			}

			Cindex := 0
			for len(A) > 0 {
				sublen := globalWorkSize
				if len(A) < sublen {
					sublen = len(A)
				}
				log.Printf("sublen: %d", sublen)
				var subA []int32
				var subB []int32
				var subC []int32
				subA, A = A[:sublen], A[sublen:]
				subB, B = B[:sublen], B[sublen:]

				var aBytes []byte
				if aBytes, err = getBytes(subA, 0, sublen); err != nil {
					panic(err)
				}
				lenAbytes := uint32(len(aBytes))

				var bBytes []byte
				if bBytes, err = getBytes(subB, 0, sublen); err != nil {
					panic(err)
				}
				lenBbytes := uint32(len(bBytes))

				if lenAbytes != lenBbytes {
					panic(fmt.Errorf("lenAbytes %d != lenBbytes %d", lenAbytes, lenBbytes))
				}

				// JMT: C is just as long as A and B
				lenCbytes := lenAbytes

				if err = queue.EnqueueWriteBuffer(bufA, aBytes, 0); err != nil {
					panic(err)
				}

				if err = queue.Finish(); err != nil {
					panic(err)
				}

				if err = queue.EnqueueWriteBuffer(bufB, bBytes, 0); err != nil {
					panic(err)
				}

				if err = queue.Finish(); err != nil {
					panic(err)
				}

				if err = kernel.SetArg(3, int32(lenCbytes)); err != nil {
					panic(err)
				}

				if err = queue.Finish(); err != nil {
					panic(err)
				}

				if err = queue.EnqueueKernel(kernel, []cl.Size{0}, []cl.Size{cl.Size(globalWorkSize)}, []cl.Size{cl.Size(localWorkSize)}); err != nil {
					panic(err)
				}

				if err = queue.Finish(); err != nil {
					panic(err)
				}

				if outBuf, err := queue.EnqueueReadBuffer(bufC, 0, lenCbytes); err != nil {
					panic(err)
				} else {
					if subC, err = getInt32s(outBuf, 0, sublen); err != nil {
						panic(err)
					}
					copy(C[Cindex:], subC)
					Cindex += sublen
				}
			}
			// test
			for i := int32(0); i < elements; i++ {
				if C[i] != i<<1 {
					log.Fatalf("Output at i=%d is incorrect: %d != %d", i, C[i], i<<1)
				}
			}
		}
	}
}

// get_local_work_size calculates the best local work size for a given kernel based on the device's limitations.
// The argument max represents the maximum possible value for local work size -- the number of elements in the whole run.
func getLocalWorkSize(dev cl.Device, kernel *cl.Kernel, max int) (localWorkSize int, err error) {
	localWorkSize = max

	var maxWorkGroupSize int
	if maxWorkGroupSize, err = strconv.Atoi(fmt.Sprintf("%d", dev.Property(cl.DEVICE_MAX_WORK_GROUP_SIZE))); err != nil {
		return 0, err
	}
	log.Printf("max work group size: %d", maxWorkGroupSize)

	if localWorkSize > maxWorkGroupSize {
		localWorkSize = maxWorkGroupSize
	}

	var workItemDimensions int
	if workItemDimensions, err = strconv.Atoi(fmt.Sprintf("%d", dev.Property(cl.DEVICE_MAX_WORK_ITEM_DIMENSIONS))); err != nil {
		return 0, err
	}
	log.Printf("work item dimensions: %d", workItemDimensions)

	var maxWorkItemSizes []int
	maxWorkItemSizesStr := fmt.Sprintf("%v", dev.Property(cl.DEVICE_MAX_WORK_ITEM_SIZES))
	for _, v := range strings.Split(maxWorkItemSizesStr[1:len(maxWorkItemSizesStr)-1], " ") {
		var val int
		if val, err = strconv.Atoi(v); err != nil {
			return 0, err
		}
		maxWorkItemSizes = append(maxWorkItemSizes, val)
	}

	if len(maxWorkItemSizes) != workItemDimensions {
		return 0, fmt.Errorf("length of max_work_item_sizes %d != work_item_dimensions %d", len(maxWorkItemSizes), workItemDimensions)
	}

	log.Printf("max work item sizes: %v", maxWorkItemSizes)

	if localWorkSize > maxWorkItemSizes[0] {
		localWorkSize = maxWorkItemSizes[0]
	}

	var kernelWorkGroupSize int
	if kernelWorkGroupSize, err = strconv.Atoi(fmt.Sprintf("%d", kernel.WorkGroupProperty(dev, cl.KERNEL_WORK_GROUP_SIZE))); err != nil {
		return 0, err
	}
	log.Printf("kernel work group size: %d", kernelWorkGroupSize)

	if localWorkSize > kernelWorkGroupSize {
		localWorkSize = kernelWorkGroupSize
	}

	var kernelPreferredWorkGroupSizeMultiple int
	if kernelPreferredWorkGroupSizeMultiple, err = strconv.Atoi(fmt.Sprintf("%d", kernel.WorkGroupProperty(dev, cl.PREFERRED_WORK_GROUP_SIZE_MULTIPLE))); err != nil {
		return 0, err
	}
	log.Printf("kernel preferred work group size multiple: %d", kernelPreferredWorkGroupSizeMultiple)

	localWorkSize = int(float64(localWorkSize)/float64(kernelPreferredWorkGroupSizeMultiple)) * kernelPreferredWorkGroupSizeMultiple

	log.Printf("local work size: %d", localWorkSize)

	return localWorkSize, nil
}

// get_global_work_size calculates the best global work size for a given kernel based on the device's limitations and the sizes of the various bits of data used in the algorithm.
// The argument max represents the maximum possible value for local work size -- the number of elements in the whole run.
func getGlobalWorkSize(dev cl.Device, kernel *cl.Kernel, max int, staticData int, bpeSingle int, bpeTotal int, localWorkSize int) (globalWorkSize int, err error) {
	globalWorkSize = max

	var globalMemSize int
	if globalMemSize, err = strconv.Atoi(fmt.Sprintf("%d", dev.Property(cl.DEVICE_GLOBAL_MEM_SIZE))); err != nil {
		return 0, err
	}

	log.Printf("global mem size: %d", globalMemSize)

	maxTotal := int((float64(globalMemSize) - float64(staticData)) / float64(bpeTotal))

	if globalWorkSize > maxTotal {
		log.Printf("reducing global_work_size from %d to %d", globalWorkSize, maxTotal)
		globalWorkSize = maxTotal
	}

	var memAllocSize int
	if memAllocSize, err = strconv.Atoi(fmt.Sprintf("%d", dev.Property(cl.DEVICE_MAX_MEM_ALLOC_SIZE))); err != nil {
		return 0, err
	}
	log.Printf("max mem alloc size: %d", memAllocSize)

	maxSingle := int(float64(memAllocSize) / float64(bpeSingle))
	if globalWorkSize > maxSingle {
		log.Printf("reducing global_work_size from %d to %d", globalWorkSize, maxSingle)
		globalWorkSize = maxSingle
	}

	globalWorkSize = int(float64(globalWorkSize)/float64(localWorkSize)) * localWorkSize

	log.Printf("global work size: %d", globalWorkSize)

	return globalWorkSize, nil
}

// get_bytes converts an array of int32's into a byte array.
func getBytes(arr []int32, offset int, len int) (out []byte, err error) {
	aBuffer := new(bytes.Buffer)
	for i := offset; i < offset+len; i++ {
		if err = binary.Write(aBuffer, binary.LittleEndian, arr[i]); err != nil {
			return nil, err
		}
	}
	return aBuffer.Bytes(), nil
}

// get_int32s converts a byte byte into an array of int32's.
func getInt32s(out []byte, offset int, len int) (arr []int32, err error) {
	cBuffer := bytes.NewReader(out)
	var elem int32
	// JMT: THIS IS HORRIBLE
	for i := 0; i < offset+len; i++ {
		if err = binary.Read(cBuffer, binary.LittleEndian, &elem); err != nil {
			return nil, err
		}
		if i >= offset {
			arr = append(arr, elem)
		}
	}
	return arr, nil
}
